modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/filters/SelectionFilter.js

```diff
@@ -113,52 +117,82 @@ function SelectionFilter({
 	const [infiniteLoaderRendered, setInfiniteLoaderRendered] = useState(false);
 	const [exclude, setExclude] = useState(!!selectedData?.exclude);
 
-	const loaderVisible = items?.length < total;
+	const loaderVisible = !localItems?.length && items?.length < total;
 
 	useEffect(() => {
 		setSelectedItems(selectedData?.selectedItems || []);
 	}, [selectedData]);
 
-	useEffect(() => {
-		if (query === search) {
-			return;
-		}
-
-		setCurrentPage(1);
-
-		setSearch(query);
-	}, [query, search]);
-
-	const isMounted = useIsMounted();
-
-	useEffect(() => {
-		if (apiURL) {
+	const loadSelectionItems = function () {
+		if (apiURL && !localItems?.length) {
 			setLoading(true);
 
 			fetchData(apiURL, search, currentPage)
-				.then((data) => {
+				.then((response) => {
+					const selectionItems = response.items.map((item) => {
+						return {
+							label: itemLabel
+								? getValueFromItem(item, itemLabel)
+								: item.label,
+							value: itemKey ? item[itemKey] : item.value,
+						};
+					});
+
 					if (!isMounted()) {
 						return;
 					}
 
 					setLoading(false);
 
 					if (currentPage === 1) {
-						setItems(data.items);
+						setItems(selectionItems);
 					}
 					else {
-						setItems((items) => [...items, ...data.items]);
+						setItems((items) => [...items, ...selectionItems]);
+					}
+
+					if (
+						firstRequest &&
+						response.totalCount <= DEFAULT_PAGE_SIZE &&
+						autocompleteEnabled
+					) {
+						setLocalItems(selectionItems);
 					}
 
-					setTotal(data.totalCount);
+					setTotal(response.totalCount);
+					setFirstRequest(false);
 				})
 				.catch(() => {
 					if (isMounted()) {
 						setLoading(false);
 					}
 				});
 		}
-	}, [autocompleteEnabled, currentPage, isMounted, search, apiURL]);
+		else if (localItems?.length && autocompleteEnabled) {
+			setItems(
+				search
+					? localItems?.filter(({label}) =>
+							label.toLowerCase().match(search.toLowerCase())
+					  )
+					: localItems
+			);
+		}
+	};
+
+	const debouncedQuery = debounce((value) => {
+		setCurrentPage(1);
+		setSearch(value);
+	}, DEFAULT_DEBOUNCE_DELAY);
+
+	const handleAutocompleteQuery = (query) => {
+		debouncedQuery(query);
+	};
```

Comment:
We probably don't need this intermediary function like this, we can just use `debouncedQuery` directly on line 278